### PR TITLE
Refactored ops on size to be dispatcher ops

### PR DESCRIPTION
**Pull request by Chillee**
An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83813
* #83814
* #83822
* __->__ #83719

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -18,6 +18,23 @@
...
 SYM_FUNCTION_MODE = None
 
+math_lib = torch.library.Library("math", "DEF")
+
+math_lib.define("size(Tensor x, int dim) -> int")
+math_lib.define("stride(Tensor x, int dim) -> int")
+math_lib.define("mul(int a, int b) -> int")
+
+def size(x, dim):
+    return x.size(dim)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:28]: Why aren't these just the aten.size ops?

Comment (Chillee) [reply torch/fx/experimental/symbolic_shapes.py:28]: that's a ... good point lol.

File: test/test_proxy_tensor.py [hunk 2]
```diff
@@ -710,10 +710,10 @@ def f(a):
         r = str(make_fx(f, tracing_mode="symbolic")(torch.empty(4)).code).strip()
         self.assertExpectedInline(r, """\
 def forward(self, a_1):
-    size = a_1.size(0);  a_1 = None
-    mul = size * 2;  size = None
+    size = torch.ops.aten.size(a_1, 0);  a_1 = None
+    mul = torch.ops.math.mul(size, 2);  size = None
```
Comment (ezyang) [inline test/test_proxy_tensor.py:714]: good bye size * 2, I will miss you lol

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -566,24 +566,20 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         flat_arg_tensors = [
             i for i in tree_flatten((args, kwargs))[0] if isinstance(i, FakeTensor)
         ]
-        has_symbolic_sizes = any([i.has_sym_ints for i in flat_arg_tensors])
+        flat_symints = [
+            i
+            for i in tree_flatten((args, kwargs))[0]
+            if isinstance(i, torch._C.SymIntNode)
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:572]: can someone pleaaase give this a better name lol. torch.SymIntNode or torch.SymInt

File: torch/fx/experimental/proxy_tensor.py [hunk 4]
```diff
@@ -467,7 +468,18 @@ def __sym_dispatch__(self, func, types, args, kwargs):
...
+        import operator
+        mapped = {
+            operator.mul: torch.ops.math.mul,
+            operator.eq: torch.ops.math.eq,
+            operator.gt: torch.ops.math.gt,
+            operator.lt: torch.ops.math.lt,
+        }
+        if func not in mapped:
+            print(func)
+            raise RuntimeError(f"Don't support {func} on symints")
```
Comment (ezyang) [inline torch/fx/experimental/proxy_tensor.py:480]: This should get fixed at the source. SymDispatchMode should get torch.ops.math calls to begin with. Don't post facto translate them in proxy tensor.

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -18,6 +18,19 @@
 
 SYM_FUNCTION_MODE = None
 
+math_lib = torch.library.Library("math", "DEF")
+
+math_lib.define("mul(int a, int b) -> int")
+math_lib.define("eq(int a, int b) -> bool")
+math_lib.define("gt(int a, int b) -> bool")
+math_lib.define("lt(int a, int b) -> bool")
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:26]: FWIW, JIT "prim" actually already defines operators that look like these. They don't have dispatcher entries though.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:26]: How come these don't have SymInt in their schema?

File: torch/_subclasses/fake_tensor.py (hunk 3 above)
Comment (Chillee) [reply torch/_subclasses/fake_tensor.py:572]: where do I register it to give it a better name?

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:572]: I think monkeying around with `torch/__init__.py` so it gets set there is the idea

Comment (albanD) [reply torch/_subclasses/fake_tensor.py:572]: If you want printing to also look nice, you can update the `__module__` field to match.

File: functorch/functorch/_src/fx_minifier.py [hunk 6]
```diff
@@ -296,4 +296,5 @@ def try_granularity(failing_state, granularity, use_non_granular):
     print(f"Made {num_queries} queries")
     failing_fx = fx.GraphModule(fail_f, failing_state.graph)
     dump_state(failing_fx, failing_state.inps)
+    print("Wrote minimal repro out to repro.py")
```
Comment (albanD) [inline functorch/functorch/_src/fx_minifier.py:299]: raw print?

File: torch/_meta_registrations.py [hunk 7]
```diff
@@ -211,6 +211,23 @@ def meta_bernoulli(self, *, generator=None, out):
...
+    device=None,
+    pin_memory=None,
+    memory_format=None,
+):
+    dtype = self.dtype if dtype is None else dtype
+    device = self.device if device is None else device
+    layout = self.layout if layout is None else layout
+    assert pin_memory is None
+    assert memory_format is None
+    return self.new_empty(self.shape, dtype=dtype, device=device, pin_memory=pin_memory)
```
Comment (bdhirsh) [inline torch/_meta_registrations.py:228]: just curious - is this python decomp needed now? (aka something now breaks with the C++ decomp)

Comment (Chillee) [reply torch/_meta_registrations.py:228]: `_to_copy` isn't `CompositeImplicitAutograd`, so I don't think there's a "C++ decomp"? Or do you mean the meta function implemented in C++?

Comment (bdhirsh) [reply torch/_meta_registrations.py:228]: oh yep, I mean the meta function implemented in C++; is that not useable with the changes in this PR?

File: functorch/functorch/_src/fx_minifier.py (hunk 6 above)
Comment (Chillee) [reply functorch/functorch/_src/fx_minifier.py:299]: No, this is intended - just a small QOL improvement to minifier I'm folding into this change lol.

File: torch/_meta_registrations.py (hunk 7 above)
Comment (Chillee) [reply torch/_meta_registrations.py:228]: I don't think we use any C++ meta functions today, although I think it's plausible we could in the future.

File: test/test_proxy_tensor.py (hunk 2 above)
Comment (Chillee) [reply test/test_proxy_tensor.py]: It's back :)

**Review by ezyang — APPROVED** [review]

File: torch/_meta_registrations.py (hunk 7 above)
Comment (IvanYashchuk) [reply torch/_meta_registrations.py:228]: This implementation is slightly incorrect and it seems there were no tests to catch that: https://github.com/pytorch/pytorch/issues/84335.

**Pull request closed without merging**